sql/colexec: TestSortedDistinct failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/0f473848083559c8a98be032949df9428068c223

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestSortedDistinct PKG=github.com/cockroachdb/cockroach/pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1566262&tab=buildLog

```

panic: runtime error: slice bounds out of range [recovered]
	panic: runtime error: slice bounds out of range

goroutine 1230 [running]:
testing.tRunner.func1(0xc0006e0e00)
	/usr/local/go/src/testing/testing.go:830 +0x392
panic(0x1eebe20, 0x3bc8990)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/col/coldata.(*Bytes).Get(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/col/coldata/bytes.go:90
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*sortedDistinctBytesOp).Next(0xc0000ab560, 0x26cc420, 0xc000052108, 0xc0008d1bf0, 0xc0003b09f0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/distinct.eg.go:619 +0xd40
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*boolVecToSelOp).Next(0xc0003b0940, 0x26cc420, 0xc000052108, 0x2, 0xc0008d1bf0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/bool_vec_to_sel.go:36 +0x87
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*opTestOutput).next(0xc0003b0b40, 0x26cc420, 0xc000052108, 0x1, 0x2, 0xc0008d1bf0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/utils_test.go:765 +0xbb
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*opTestOutput).Verify(0xc0003b0b40, 0xc0003b0b40, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/utils_test.go:784 +0x78
github.com/cockroachdb/cockroach/pkg/sql/colexec.runTestsWithoutAllNullsInjection.func1(0xc0006e0e00, 0xc0003b8e80, 0x1, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/utils_test.go:211 +0x174
github.com/cockroachdb/cockroach/pkg/sql/colexec.runTestsWithFn.func1(0xc0006e0e00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/utils_test.go:347 +0x298
testing.tRunner(0xc0006e0e00, 0xc0000ab500)
	/usr/local/go/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:916 +0x35a

goroutine 1 [chan receive]:
testing.(*T).Run(0xc0006e0200, 0x2156dbb, 0x12, 0x2206120, 0x4c4701)
	/usr/local/go/src/testing/testing.go:917 +0x381
testing.runTests.func1(0xc00061e100)
	/usr/local/go/src/testing/testing.go:1157 +0x78
testing.tRunner(0xc00061e100, 0xc0007b5e10)
	/usr/local/go/src/testing/testing.go:865 +0xc0
testing.runTests(0xc00000ef40, 0x3be8fe0, 0x5d, 0x5d, 0x0)
	/usr/local/go/src/testing/testing.go:1155 +0x2a9
testing.(*M).Run(0xc000273e00, 0x0)
	/usr/local/go/src/testing/testing.go:1072 +0x162
github.com/cockroachdb/cockroach/pkg/sql/colexec.TestMain(0xc000273e00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/main_test.go:24 +0x30
main.main()
	_testmain.go:274 +0x13e

goroutine 35 [syscall]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:139 +0x9c
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:29 +0x41

goroutine 50 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:73 +0x74
created by github.com/cockroachdb/cockroach/pkg/util/log.init.7
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:45 +0x35

goroutine 51 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.signalFlusher()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:100 +0x126
created by github.com/cockroachdb/cockroach/pkg/util/log.init.7
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:46 +0x4d

goroutine 7 [select]:
github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view.(*worker).start(0xc0002d0100)
	/go/src/github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view/worker.go:152 +0xdd
created by github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view.init.0
	/go/src/github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view/worker.go:29 +0x57

goroutine 1202 [chan receive]:
testing.(*T).Run(0xc0006e0e00, 0xc0004aa780, 0x14, 0xc0000ab500, 0x2)
	/usr/local/go/src/testing/testing.go:917 +0x381
github.com/cockroachdb/cockroach/pkg/sql/colexec.runTestsWithFn(0xc0006e0200, 0xc00033e9a0, 0x1, 0x1, 0x0, 0x0, 0x0, 0xc0008d11a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/utils_test.go:329 +0x15e
```

--- a/pkg/col/coldata/bytes.go
+++ b/pkg/col/coldata/bytes.go
@@ -86,6 +86,8 @@ func (b *Bytes) maybeBackfillOffsets(i int) {
 
 // Get returns the ith []byte in Bytes. Note that the returned byte slice is
 // unsafe for reuse if any write operation happens.
+// NOTE: if ith element was never set in any way, the behavior of Get is
+// undefined.
 func (b *Bytes) Get(i int) []byte {
 	return b.data[b.offsets[i]:b.offsets[i+1]]
 }

--- a/pkg/sql/colexec/distinct_tmpl.go
+++ b/pkg/sql/colexec/distinct_tmpl.go
@@ -365,7 +365,7 @@ func _CHECK_DISTINCT(
 	checkIdx int, outputIdx int, lastVal _GOTYPE, col []_GOTYPE, outputCol []bool,
 ) { // */}}
 
-	// {{define "checkDistinct"}}
+	// {{define "checkDistinct" -}}
 	v := execgen.UNSAFEGET(col, int(checkIdx))
 	var unique bool
 	_ASSIGN_NE(unique, v, lastVal)
@@ -390,19 +390,26 @@ func _CHECK_DISTINCT_WITH_NULLS(
 	outputCol []bool,
 ) { // */}}
 
-	// {{define "checkDistinctWithNulls"}}
+	// {{define "checkDistinctWithNulls" -}}
 	null := nulls.NullAt(uint16(checkIdx))
-	v := execgen.UNSAFEGET(col, int(checkIdx))
-	if null != lastValNull {
-		// Either the current value is null and the previous was not or vice-versa.
-		outputCol[outputIdx] = true
-	} else if !null {
-		// Neither value is null, so we must compare.
-		var unique bool
-		_ASSIGN_NE(unique, v, lastVal)
-		outputCol[outputIdx] = outputCol[outputIdx] || unique
+	if null {
+		if !lastValNull {
+			// The current value is null while the previous was not.
+			outputCol[outputIdx] = true
+		}
+	} else {
+		v := execgen.UNSAFEGET(col, int(checkIdx))
+		if lastValNull {
+			// The previous value was null while the current is not.
+			outputCol[outputIdx] = true
+		} else {
+			// Neither value is null, so we must compare.
+			var unique bool
+			_ASSIGN_NE(unique, v, lastVal)
+			outputCol[outputIdx] = outputCol[outputIdx] || unique
+		}
+		execgen.COPYVAL(lastVal, v)
 	}
-	execgen.COPYVAL(lastVal, v)
 	lastValNull = null
 	// {{end}}
 

